From de80831314ecb76ac22b19b249467a600129a9e3 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Fri, 14 Apr 2017 16:43:31 +0800
Subject: advapi32: Read the configured object list for the performance
 provider.

FIXME: it's not currently used, but the queries should be matched
against the configured object lists, and the providers should be
loaded and called only in case of a match.
---
 dlls/advapi32/registry.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/dlls/advapi32/registry.c b/dlls/advapi32/registry.c
index 0aa38c5b9b3..5af8128748b 100644
--- a/dlls/advapi32/registry.c
+++ b/dlls/advapi32/registry.c
@@ -1488,6 +1488,7 @@ struct perf_provider
 {
     HMODULE perflib;
     WCHAR linkage[MAX_PATH];
+    WCHAR objects[MAX_PATH];
     PM_OPEN_PROC *pOpen;
     PM_CLOSE_PROC *pClose;
     PM_COLLECT_PROC *pCollect;
@@ -1511,6 +1512,7 @@ static void *get_provider_entry(HKEY perf, HMODULE perflib, const char *name)
 
 static BOOL load_provider(HKEY root, const WCHAR *name, struct perf_provider *provider)
 {
+    static const WCHAR object_listW[] = { 'O','b','j','e','c','t',' ','L','i','s','t',0 };
     static const WCHAR performanceW[] = { 'P','e','r','f','o','r','m','a','n','c','e',0 };
     static const WCHAR libraryW[] = { 'L','i','b','r','a','r','y',0 };
     static const WCHAR linkageW[] = { 'L','i','n','k','a','g','e',0 };
@@ -1543,6 +1545,16 @@ static BOOL load_provider(HKEY root, const WCHAR *name, struct perf_provider *pr
     if (err != ERROR_SUCCESS)
         return FALSE;
 
+    provider->objects[0] = 0;
+    len = sizeof(buf) - sizeof(WCHAR);
+    err = RegQueryValueExW(perf, object_listW, NULL, &type, (BYTE *)buf, &len);
+    if (err == ERROR_SUCCESS && (type == REG_SZ || type == REG_MULTI_SZ))
+    {
+        memcpy(provider->objects, buf, len);
+        provider->objects[len / sizeof(WCHAR)] = 0;
+        TRACE("Object List: %s\n", debugstr_w(provider->objects));
+    }
+
     len = sizeof(buf) - sizeof(WCHAR);
     err = RegQueryValueExW(perf, libraryW, NULL, &type, (BYTE *)buf, &len);
     if (err != ERROR_SUCCESS || !(type == REG_SZ || type == REG_EXPAND_SZ))
-- 
2.13.1

